### DISCUSSION TITLE
Pass schema to function and it returns infer-ed type of schema

### DISCUSSION THREAD

**Discussion by denvaki** [colinhacks/zod#1808] [Q&A] [answered]
Hey, thank you so much for this library, it is awesome!!

I try to achieve validation function for http responses, that could:

1. accept  response schema as argument
2. validate response against schema
3. return validated response or throw an error

Below is my maximum I could create :sweat_smile:  and it works in simple cases.

```ts
async function requestWithValidate<T>(
	httpCall: () => Promise<unknown>,
	schema: z.ZodSchema<T>
): Promise<T> {
	try {
		const response = await httpCall();
		return schema.parse(response);
	} catch (e) {
		if (e instanceof ZodError) {
			e = e.format();
		}
		throw e;
	}
};
```

However, when I pass schema that contains `ZodEffect`(schema field with pre-processor for example) I have type incompatibility.
I have it for `Date` fields and for them I get `Type 'unknown' is not assignable to type 'Date'`

Pre-processor:
``` ts
const stringToDateConvert = z.preprocess((arg) => {
    if (typeof arg === 'string' || arg instanceof Date) {
        return new Date(arg);
    }
}, z.date());
```
Example schema:
``` typescript
z.object({
    id: z.number(),
    date: stringToDateConvert.nullable(),
}).nullable();
```

Could you please help with setup of generics for this function that could work even with any Zod type?

**Comment by denvaki** [comment]
I found workaround to infer type from `_type` like this:
```typescript
export = async function requestWithValidate
        <T extends { parse(data: unknown, params?: Partial<ParseParams>): unknown; _type: unknown }>
         (httpCall: () => Promise<unknown>, schema: T): Promise<T['_type']> {
	try {
		const response = await httpCall();
		return schema.parse(response);
	} catch (e) {
		if (e instanceof ZodError) {
			e = e.format();
		}
		throw e;
	}
};
```

@colinhacks , mb we could add this into doc? I think my case faced quite often for others as well

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
async function requestWithValidate<Schema extends z.ZodTypeAny> (
    httpCall: () => Promise<unknown>,
    schema: Schema
): Promise<z.infer<Schema>> {
    try {
        const response = await httpCall()
        return schema.parse( response )
    } catch ( e ) {
        if ( e instanceof z.ZodError ) {
            e = e.format()
        }
        throw e
    }
}
```

**Comment by artem-tarasenko** [comment]
Hi @JacobWeisenburger, I didn't want to create a new discussion, so if you don't mind I will ask here as it's a somewhat related question.
I was browsing through "generic return-type" discussions and couldn't find an answer to a couple of questions, so...

Why is it recommended to use `<T extends z.ZodTypeAny>` extending ZodTypeAny, if ZodType generic has an expected return value (Output)? 

Why not `fn<z.ZodType<T>>(schema: T): T` ? I mean, that's how I would expect that to work. Using "any" generic to add a return type manually doesn't make sense IMHO, but since it is the suggested way, I'm curious why so?
